sbom/spdx/encoder.go

```diff
@@ -0,0 +1,417 @@
+
+		// Record Repositories for this package.
+		if r.Repository != nil {
+			rRepoId, err := strconv.Atoi(r.Repository.ID)
+			if err != nil {
+				return nil, err
+			}
+
+			repo, ok := repoMap[rRepoId]
+			// Record the Repository if we haven't seen it yet.
+			if !ok {
+				repo = newSpdxPackageFromRepository(r.Repository)
+
+				repoMap[rRepoId] = repo
+				repoIds = append(repoIds, rRepoId)
+			}
+
+			rel := &v2_3.Relationship{
+				RefA:         v2common.MakeDocElementID("", string(pkg.PackageSPDXIdentifier)),
+				RefB:         v2common.MakeDocElementID("", string(repo.PackageSPDXIdentifier)),
+				Relationship: "CONTAINED_BY",
+			}
+			pkgRels[rPkgId] = append(pkgRels[rPkgId], rel)
+		}
+	}
+
+	// Now that we have all the data necessary to create the SPDX document,
+	// we need to order it since the IndexRecords aren't in a deterministic order.
+	// This is particular helpful for testing, but it wouldn't be unreasonable
+	// for a user to want to diff different versions of an SPDX of the same IndexReport.
+	slices.Sort(pkgIds)
+	for _, id := range pkgIds {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		pkg := pkgMap[id]
+		out.Packages = append(out.Packages, pkg)
+
+		rels := pkgRels[id]
+		slices.SortFunc(rels, cmpRelationship)
+		out.Relationships = append(out.Relationships, rels...)
+	}
+
+	slices.Sort(distIds)
+	for _, id := range distIds {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		dist := distMap[id]
+		out.Packages = append(out.Packages, dist)
+	}
+
+	slices.Sort(repoIds)
+	for _, id := range repoIds {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		repo := repoMap[id]
+		out.Packages = append(out.Packages, repo)
+	}
+
+	return out, nil
+}
+
+func newSpdxPackageFromPackage(p *claircore.Package) *v2_3.Package {
+	pkg := &v2_3.Package{
+		PackageName:             p.Name,
+		PackageSPDXIdentifier:   v2common.ElementID("Package-" + p.ID),
+		PackageVersion:          p.Version,
+		PackageDownloadLocation: "NOASSERTION",
+	}
+	return pkg
+}
+
+func newSpdxPackageFromDistribution(d *claircore.Distribution) *v2_3.Package {
+	var extRefs []*v2_3.PackageExternalReference
+
+	if d.CPE.String() != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "SECURITY",
+			RefType:  "cpe23Type",
+			Locator:  d.CPE.String(),
+		})
+	}
+
+	if d.DID != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "did",
+			Locator:  d.DID,
+		})
+	}
+
+	if d.VersionID != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "version_id",
+			Locator:  d.VersionID,
+		})
+	}
+
+	if d.PrettyName != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "pretty_name",
+			Locator:  d.PrettyName,
+		})
+	}
+
+	dist := &v2_3.Package{
+		PackageName:               d.Name,
+		PackageSPDXIdentifier:     v2common.ElementID("Distribution-" + d.ID),
```

Comment:
earlier there is a call like `fmt.Sprintf()` similar to this. Is there a reason for the change (I also may prefer the `+`)